src/newrelic/processor.go

```diff
@@ -663,42 +707,93 @@ func harvestByType(ah *AppHarvest, args *harvestArgs, ht HarvestType) {
 		log.Debugf("harvesting error events")
 
 		errorEvents := harvest.ErrorEvents
 		harvest.ErrorEvents = NewErrorEvents(eventConfigs.ErrorEventConfig.Limit)
-		considerHarvestPayload(errorEvents, args)
+		considerHarvestPayload(errorEvents, args, duc)
 	}
 
 	if ht&HarvestTxnEvents == HarvestTxnEvents && eventConfigs.AnalyticEventConfig.Limit != 0 {
 		log.Debugf("harvesting transaction events")
 
 		txnEvents := harvest.TxnEvents
 		harvest.TxnEvents = NewTxnEvents(eventConfigs.AnalyticEventConfig.Limit)
-		considerHarvestPayloadTxnEvents(txnEvents, args)
+		considerHarvestPayloadTxnEvents(txnEvents, args, duc)
 	}
 
 	if ht&HarvestSpanEvents == HarvestSpanEvents && eventConfigs.SpanEventConfig.Limit != 0 {
 		log.Debugf("harvesting span events")
 
 		spanEvents := harvest.SpanEvents
 		harvest.SpanEvents = NewSpanEvents(eventConfigs.SpanEventConfig.Limit)
-		considerHarvestPayload(spanEvents, args)
-
+		considerHarvestPayload(spanEvents, args, duc)
 	}
 
 	if ht&HarvestLogEvents == HarvestLogEvents && eventConfigs.LogEventConfig.Limit != 0 {
 		log.Debugf("harvesting log events")
 
 		logEvents := harvest.LogEvents
 		harvest.LogEvents = NewLogEvents(eventConfigs.LogEventConfig.Limit)
-		considerHarvestPayload(logEvents, args)
+		considerHarvestPayload(logEvents, args, duc)
+
+	}
+	// Only harvest data usage metrics if metrics are being harvested
+	if ht&HarvestDefaultData == HarvestDefaultData {
+		if args.blocking {
+			harvestDataUsage(args, duc)
+		} else {
+			go harvestDataUsage(args, duc)
+		}
+	}
+}
+
+func harvestDataUsage(args *harvestArgs, duc dataUsageController) {
+	duc.wg.Wait()
+	if len(duc.duc) == 0 {return} // no usage metrics found
+
+	sumPayload := 0
+	sumResponse := 0
+	sumAttempts := 0
+	type dataUsageMetrics struct {
+		attempts int
+		payloadSize int
+		responseSize int
+	}
+	dataUsageMap := make(map[string]dataUsageMetrics)
+
+	loop := true
+	for loop {
+		select {
+		case d, ok := <-duc.duc:
+			if ok {
+				usage := dataUsageMap[d.endpoint_name]
+				usage.payloadSize += d.payloadSize
+				usage.responseSize += d.responseSize
+				usage.attempts += 1
+				dataUsageMap[d.endpoint_name] = usage
+				sumPayload += d.payloadSize
+				sumResponse += d.responseSize
+				sumAttempts += 1
+			}
+		default:
+			loop = false
+		}
+	}
 
+	metrics := NewMetricTable(limits.MaxMetrics, time.Now())
+	for name, data := range(dataUsageMap) {
+		metrics.AddRaw([]byte("Supportability/" + args.agentLanguage + "/" + args.collector + "/" + name + "/Output/Bytes"),
+					   "", "", [6]float64{float64(data.attempts), float64(data.payloadSize), float64(data.responseSize), 0.0, 0.0, 0.0}, Forced)
 	}
+	metrics.AddRaw([]byte("Supportability/" + args.agentLanguage + "/" + args.collector + "/Output/Bytes"),
+				   "", "", [6]float64{float64(sumAttempts), float64(sumPayload), float64(sumResponse), 0.0, 0.0, 0.0}, Forced)
+	metrics = metrics.ApplyRules(args.rules)
+	considerHarvestPayload(metrics, args, duc)
```

Comment:
What keeps this from creating another cycle ending with sending data usage metrics again into an infinite recurse?

Reply:
`considerHarvestPayload` only calls the `Exectue()` to send the data and adds the resulting usage to the data usage channel. It does not trigger another harvest nor call `harvestDataUsage()`. The next harvest cycle that includes metrics will then call `harvestDataUsage()` again, pulling any usage since the previous call to be included in this one.